[C++ Interop] Support C++ parameters in Obj-C++ methods.

drodriguez opened the pull request:
`importMethodParamsAndReturnType` and `importFunctionParameterList`
share a lot of code but differed in how they treated C++ types.

Extract the common code and merge the differences in code between the
two functions into one single function and use it from both places. At the
end it needs also a small function to account for the slightly different way
Obj-C methods deal with error and completion callbacks.

NOTE: Even if `const T&` parameters are now correctly imported as `T` at
the Swift level, in the case of initializers they are not correctly
imported as `@in T` and a crash will occur when using them.

Relates to #60638

/cc @plotfi @bulbazord

hyp commented:
needs to be tested :)

drodriguez commented:
The test I am using is at #60638, but because of the missing `@in` it is not fully how I want it, so I did not include it in this draft.

As we had talked, I will try to merge both `import*Param` methods instead of this copy/paste job. I can include the test then (I am thinking into trying to solve the missing `@in` in some other PR and get this fixed partially).

zoecarver commented:
I feel like we could unify these two functions. But, not something you have to do in this patch. This LGTM (of course with some tests).

drodriguez commented:
@swift-ci please smoke test

drodriguez commented:
Version with the common code extracted and a test for the Obj-C++ methods having the right parameters. The test includes a couple of FIXME for those missing `@in` and also notes about maybe needed to be `@in_guaranteed`. I do not intend to deal with those problems in this PR. This PR only shows that sharing the code the parameters of the Obj-C++ methods are imported as the C++ types, and not as `UnsafePointer<T>`.

plotfi commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2227,6 +2227,157 @@ ImportedType ClangImporter::Implementation::importFunctionParamsAndReturnType(
+  } else if (auto refType = dyn_cast<clang::ReferenceType>(paramTy)) {
+    // We don't support reference type to a dependent type, just bail.
+    if (refType->getPointeeType()->isDependentType()) {
+      return None;
```
Why are we missing

```
       addImportDiagnostic(
            param, Diagnostic(diag::parameter_type_not_imported, param),
            param->getSourceRange().getBegin());
```

here?

drodriguez replied:
All the `addImportDiagnostic` calls should have been moved outside `importParameterType`. The equivalent is in Line 2418 of the new version. This means that other `return None` of the new function will add the right diagnostics (like in Line 2284, which in the previous version will not have included diagnostics).

plotfi commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2227,6 +2227,157 @@ ImportedType ClangImporter::Implementation::importFunctionParamsAndReturnType(
+    PointerTypeKind pointerKind = pointeeType.getQualifiers().hasConst()
+                                      ? PTK_UnsafePointer
+                                      : PTK_UnsafeMutablePointer;
+    auto attrs = getImportTypeAttrs(param, /*isParam=*/true);
```
Could these 3 `auto attrs = getImportTypeAttrs(param, /*isParam=*/true);` be hoisted to the top of the function?

drodriguez replied:
Two reasons why I moved this into the branches (vs being on top like in the previous version):

- `attrs` was only used inside the branches and inside the branch that does `!swiftParamTy` (Line 2324). In this branch (but not in the others), there's a check for `!swiftParamTy` that returns early, so the Line 2324 will not be hit. Some of these branches do not even use `attrs`, so it was doing extra work for no reason.
- When I saw the differences in `attrs` between the funcion and the method parameters (Line 2329) I notices the `sendableByDefault`, which exist in the method parameters, but not in the function parameters. If I hoist with that value, there will be a check for `sendableByDefault` in the function parameters path which does not make a lot of sense.
- I can also hoist to where it was before, and then rewrite around Line 2329. That would do extra work up front and reuses the variable, which I normally prefer not to do (I prefer variables that do not change values, if possible, but that's my personal style that helps me think about the code).

If nobody thinks that having the `sendableByDefault` in the hoisted version is going to be confusing (or change the behaviour of the code when we do not want it), I can use that version. My order of preference will be the current version I wrote, the one with `sendableByDefault`, and finally the one with the overwritten variable.

plotfi commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2227,6 +2227,157 @@ ImportedType ClangImporter::Implementation::importFunctionParamsAndReturnType(
+             isa<clang::TemplateTypeParmType>(paramTy->getPointeeType())) {
+    auto templateParamType =
+        cast<clang::TemplateTypeParmType>(paramTy->getPointeeType());
+    auto attrs = getImportTypeAttrs(param, /*isParam=*/true);
```
ditto

plotfi commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2227,6 +2227,157 @@ ImportedType ClangImporter::Implementation::importFunctionParamsAndReturnType(
+      isInOut = true;
+  } else if (auto *templateParamType =
+                 dyn_cast<clang::TemplateTypeParmType>(paramTy)) {
+    auto attrs = getImportTypeAttrs(param, /*isParam=*/true);
```
ditto

plotfi commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2227,6 +2227,157 @@ ImportedType ClangImporter::Implementation::importFunctionParamsAndReturnType(
+    ArrayRef<GenericTypeParamDecl *> genericParams,
+    llvm::function_ref<void(Diagnostic &&)> addImportDiagnosticFn,
+    bool &isInOut, bool &isParamTypeImplicitlyUnwrapped) {
+  auto paramTy = param->getType();
```
One observation, the following here:
```
  auto paramTy = param->getType();
  auto paramAddDiag = addImportDiagnosticFn;
```

could make the diff for this function a lot more readable.

drodriguez replied:
I can rename the parameter to `paramAddDiag`. I thought that keeping `addImportDiagnosticFn` like in other functions was a better idea, but I also was not convinced by the diff being a little bit more broken because of it.

drodriguez committed:
[NFC] Extract getParameterInfo helper

The methods `importFunctionParameterList` and
`importMethodParamsAndReturnType` share a similar piece of code to
create the parameter information from the results of finding the
parameter type. In order to simplify the later refactor and to make the
changes clearer, extract the common code into a helper function and call
it from both methods, replacing the duplicated code.

drodriguez committed:
[NFC] Move some lines earlier in code flow

To make the refactor of the methods `importFunctionParameterList` and
`importMethodParamsAndReturnType` easier in a later commit, move some
lines that were intermixed with the common code earlier in each of those
two methods. Even if some lines move earlier, there should be no
difference functionally.

drodriguez committed:
[NFCi] Extract importParameterType

Extract a helper method `importParameterType` from the
`importFunctionParameterList` code. This new helper only deals with
function parameters, but the final intention is that it will be shared
code with the code that import method parameters.

The only functional difference introduced is that before, some of the
branches will not have added an extra diagnostic about parameters
failing to import, while this version adds those diagnostics
consistently for any case.

drodriguez committed:
Test case for const T& in Obj-C++ methods

Add a test case that checks the behaviour introduced during the refactor
of `importFunctionParameterList` and `importMethodParamsAndReturnType`.

The test checks that both Obj-C++ methods and C++ functions treat `const
T&` parameters in the same way.

drodriguez committed:
[NFC] Extract getImportTypeKindForParam helper

The same piece of code was repeated in two different methods:
`importFunctionParameterList` and `importMethodParamsAndReturnType`. To
avoid repetition and simplify the later refactoring extract the code
into a helper.

The code in the two methods was happening in slightly different places,
but the modified code should still be equivalent and should not
introduce functional differences.

drodriguez commented:
@plotfi: I have split the change in smaller commits that hopefully can be reviewed independently to see the evolution of the code.

plotfi commented:
Nice work Daniel. Once this set of changes is landed I can look at changing @in to @in_guaranteed.

drodriguez commented:
@swift-ci please test

plotfi commented:
@drodriguez can you mark the change `Extract importParameterType` as "NFCi" ?

drodriguez commented:
> @drodriguez can you mark the change `Extract importParameterType` as "NFCi" ?

What does the "i" means? It is not completely NFC though, so I didn't want to mark it as such. But if I rewrite the stack I can change the commit message (I don't want to restart the testing in case it is something cosmetic, but I am sure there will be more feedback later).

plotfi commented:
> > @drodriguez can you mark the change `Extract importParameterType` as "NFCi" ?
> 
> What does the "i" means? It is not completely NFC though, so I didn't want to mark it as such. But if I rewrite the stack I can change the commit message (I don't want to restart the testing in case it is something cosmetic, but I am sure there will be more feedback later).

"intended"

plotfi commented:
Could it be possible to break up `Improve importParameterType for method parameters` into a NFC part and the part that drops the UnsafePointer?

plotfi commented:
> @drodriguez Could it be possible to break up `Improve importParameterType for method parameters` into a NFC part and the part that drops the UnsafePointer?

Once I study the non-NFC part of this patch I will probably merge. the NFC parts for the most part LGTM. The tests also look inline with what I was going for. @zoecarver @hyp do you have any other feedback on this PR?

drodriguez commented:
> @drodriguez Could it be possible to break up `Improve importParameterType for method parameters` into a NFC part and the part that drops the UnsafePointer?

I don't completely understand what you are asking me to do. Do you want me to move the pieces that support method parameters without calling it from the `importMethodParamsAndReturnType` and then another commit that removes the code from there and call the new function? Or do you mean write an intermediate version that has an extra flag to avoid that methods are imported correctly? (that sounds like extra work to be undone later).

plotfi commented:
> > @drodriguez Could it be possible to break up `Improve importParameterType for method parameters` into a NFC part and the part that drops the UnsafePointer?
> 
> I don't completely understand what you are asking me to do. Do you want me to move the pieces that support method parameters without calling it from the `importMethodParamsAndReturnType` and then another commit that removes the code from there and call the new function? Or do you mean write an intermediate version that has an extra flag to avoid that methods are imported correctly? (that sounds like extra work to be undone later).

That is a good point, don't worry about that. I think things look good as it is.

hyp reviewed (approved):
Thanks, this is good to go. Let's also run source compat suite.

hyp commented:
@swift-ci please test source compatibility

drodriguez commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2230,6 +2230,157 @@ ImportedType ClangImporter::Implementation::importFunctionParamsAndReturnType(
+                 dyn_cast<clang::TemplateTypeParmType>(paramTy)) {
+    swiftParamTy = findGenericTypeInGenericDecls(
+        *this, templateParamType, genericParams, attrs, addImportDiagnosticFn);
+  } else if (auto refType = dyn_cast<clang::ReferenceType>(paramTy)) {
```
@plotfi this is the branch that was missing in the Obj-C++ case. From line 2276 to 2304 it did not exist for methods. The branches before this are not taken because they depend on templates, but this branch checks if the `paramTy` is a reference type (which is the case for C++ structs/class), and allows using it without the `UnsafePointer`. Since `paramTy` is changed to the pointee, the later code that calls `importType` does not add the `UnsafePointer`.

PD: I don't know why I cannot comment this in the commit itself, but this branch does not appear in the commit in question because it just modifies the code around this, not this code directly.

zoecarver commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2252,114 +2403,37 @@ ParameterList *ClangImporter::Implementation::importFunctionParameterList(
 
     // Check nullability of the parameter.
-    OptionalTypeKind OptionalityOfParam = getParamOptionality(param, knownNonNull);
+    OptionalTypeKind optionalityOfParam =
+        getParamOptionality(param, knownNonNull);
```
Looks like this is the same logic twice. Can we move this inside of `importParameterType`?

drodriguez replied:
Yes. I will move it inside the function. My bad.

drodriguez replied:
I saw one problem: in `importMethodParamsAndReturnType` there's an extra usage of `importKind` outside the code that ends up in `importParameterType` (when completion handlers are dealt with) , so there will be an extra call, but hopefully that's fine.

zoecarver commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2252,114 +2403,37 @@ ParameterList *ClangImporter::Implementation::importFunctionParameterList(
-      importKind = ImportTypeKind::CFRetainedOutParameter;
-    else if (param->hasAttr<clang::CFReturnsNotRetainedAttr>())
-      importKind = ImportTypeKind::CFUnretainedOutParameter;
+    ImportTypeKind importKind = getImportTypeKindForParam(param);
```
Same here.

zoecarver commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2230,6 +2230,157 @@ ImportedType ClangImporter::Implementation::importFunctionParamsAndReturnType(
+  return importKind;
+}
+
+Optional<swift::Type> ClangImporter::Implementation::importParameterType(
+    const clang::ParmVarDecl *param, OptionalTypeKind optionalityOfParam,
+    ImportTypeKind importKind, bool allowNSUIntegerAsInt,
+    bool isNSDictionarySubscriptGetter, bool paramIsError,
+    bool paramIsCompletionHandler,
+    Optional<unsigned> completionHandlerErrorParamIndex,
+    ArrayRef<GenericTypeParamDecl *> genericParams,
+    llvm::function_ref<void(Diagnostic &&)> addImportDiagnosticFn,
+    bool &isInOut, bool &isParamTypeImplicitlyUnwrapped) {
```
Actual question: do you think there's some way we could get rid of some of the parameters? I think this is a fairly confusing API interface. Maybe we can compute some of these parameters inside of `importParameterType`. Maybe we can also get rid of the last two out parameters by returning a `ParamDecl *`.

drodriguez replied:
I think some of them can be simplified, but not all of them:

- `optionalityOfParam`: there's extra code on the function path. It depends if the function is a template specialization. It doesn't apply to methods. 
- `importKind` can be a call to `getImportTypeKindForParam`. I didn't do it to avoid more changes, but it can be done now as a NFC commit on top of this.
- `allowNSUIntegerAsInt`: the logic for methods and functions is way too different. Methods depend on selector names.
- `isNSDictionarySubscriptGetter`: only applies to methods. To calculate it one has to pass the `SpecialMethodKind`, which does not apply to functions.
- `paramIsError`: only applies to methods. To calculate one has to pass the `ForeignErrorConvention::Info` (which comes from the `methodName`) and also the `paramIndex`.
- `paramIsCompletionHandler`: similar story. Depends on the `paramIndex` and the `asyncInfo` (that comes from the `methodName`).
- `completionHandlerErrorParamIndex`: I have doubts about this myself. I think this one and the previous one represents different things: one is if the parameter is the completion handler, and the other points to the index of the _error_ of the completion handler. If someone can assure me that they actually represent the same thing, these two can be folded into one.
- `genericParams`: only applies to functions. They are calculated in `ImportDecl.cpp` from the `FuncTemplateDecl`.
- `addImportDiagnosticFn`: passed in order to be passed down to other functions.
- `isInOut` and `isParamTypeImplitlyUnwrapped`: For the case of functions one can avoid these two and do it internally, however for methods, between calculating the `swift::Type` and turning it into a `ParamDecl` there's some "early returns" (Line 3007 to 3035). I preferred to split the code in two and use these two `out bool` instead. I can also create a small type with a `swift::Type` and these two `bool` and return that instead. I don't think it makes things significantly better, but if people prefer that, it is not complicated.

zoecarver replied:
Makes sense, thank you for explaining! 

> optionalityOfParam: there's extra code on the function path. It depends if the function is a template specialization. It doesn't apply to methods.

This might be a bug. Why shouldn't it apply to methods?

> genericParams: only applies to functions. They are calculated in ImportDecl.cpp from the FuncTemplateDecl.

Again, seems like it should maybe apply to methods? Or do we not support templated methods?

> isInOut and isParamTypeImplitlyUnwrapped: For the case of functions one can avoid these two and do it internally, however for methods, between calculating the swift::Type and turning it into a ParamDecl there's some "early returns" (Line 3007 to 3035). I preferred to split the code in two and use these two out bool instead. I can also create a small type with a swift::Type and these two bool and return that instead. I don't think it makes things significantly better, but if people prefer that, it is not complicated.

OK, maybe we can do this change after the fact.

drodriguez replied:
> > optionalityOfParam: there's extra code on the function path. It depends if the function is a template specialization. It doesn't apply to methods.
> 
> This might be a bug. Why shouldn't it apply to methods?
>
> > genericParams: only applies to functions. They are calculated in ImportDecl.cpp from the FuncTemplateDecl.
> 
> Again, seems like it should maybe apply to methods? Or do we not support templated methods?

Well, Obj-C++ is a strange beast, but I don't think think Obj-C type methods can have C++-like template parameters. They have the lightweight generics, but I don't think you can mix them both. I might be wrong. Also, the language reference is whatever compiles… so no idea how to search for that.

> > isInOut and isParamTypeImplitlyUnwrapped: For the case of functions one can avoid these two and do it internally, however for methods, between calculating the swift::Type and turning it into a ParamDecl there's some "early returns" (Line 3007 to 3035). I preferred to split the code in two and use these two out bool instead. I can also create a small type with a swift::Type and these two bool and return that instead. I don't think it makes things significantly better, but if people prefer that, it is not complicated.
> 
> OK, maybe we can do this change after the fact.

I can add an extra commit into the stack with this change. I think it would make Puyan idea of mostly NFC commits a little less clear if I do the change in the extraction commit.

zoecarver commented on test/Interop/Cxx/objc-correctness/const-ref-parameter.swift:
```diff
@@ -0,0 +1,77 @@
+
+// RUN: %target-swift-frontend -c -enable-experimental-cxx-interop -enable-objc-interop -I %S/Inputs %s -emit-sil -o - | %FileCheck %s
+
+// COM: FIXME: should it be @in_guaranteed OptionsStruct?
```
Are these FIXMEs still relevant?

drodriguez replied:
Yes. We are still discussing which one of `@in` or `@in_guaranteed` avoid crashes, and also there's this weird behaviour that the initializer (this one) end up being `OptionsStruct` instead of `@in OptionStruct`.

Puyan is working on this and I think we want to try to land something about this in a later PR.

zoecarver replied:
What was the resolution of this discussion? I think we want to use `@in_guaranteed`, right? Assuming that's the case, can you create a GH issue and put that in the comment. Then I think we can have one or zero of these FIXMEs. And in that case we know what the bug is so the comment doesn't need to be framed as a question.

drodriguez replied:
I think the last decision is try to use `@in_guaranteed`. I can add links to #60601 here to make it clearer.

plotfi commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2252,114 +2403,37 @@ ParameterList *ClangImporter::Implementation::importFunctionParameterList(
+                          param->getSourceRange().getBegin());
+      return nullptr;
     }
 
```
Can we add something like `auto swiftParamTy = *swiftParamTyOpt;` here so that the lines extracting the optional can be left alone?

plotfi commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2252,114 +2403,37 @@ ParameterList *ClangImporter::Implementation::importFunctionParameterList(
+    bool isInOut = false;
+    bool isParamTypeImplicitlyUnwrapped = false;
 
+    auto swiftParamTy = importParameterType(
```
`auto swiftParamTyOpt = importParameterType(`

drodriguez reviewed (commented):
I will fold `getImportTypeKindForParam` into `importParameterType` and use the `Opt` suffix as Puyan proposes.

I will hold the changes until I see the source compatibility tests finish to not have to restart them with those changes.

drodriguez committed:
Improve importParameterType for method parameters

Adapt the helper for figuring out parameter types for functions with the
code exclusive for figuring out method parameters. This includes special
code for handling dictionary subscripts, error parameters and completion
handlers.

The function parameters behaviour should not be modified, since in order
to import function parameters all of the new pieces of code should be
skipped following the value of the flags passed into the helper.

For method parameters, there is extensive changes, specially in the
cases of importing Obj-C++ methods that use C++ types, which were
supported for functions, but not so much for methods.

drodriguez committed:
[NFC] Rename some variable names for consistency

Use capitalization correctly in one variable name, and use the name used
in other parts of the file for a parameter.

drodriguez commented:
@swift-ci please smoke test

drodriguez commented:
@swift-ci please smoke test

hyp commented:
@swift-ci please test source compatibility

zoecarver commented on test/Interop/Cxx/objc-correctness/Inputs/const-ref-parameter.h:
```diff
@@ -0,0 +1,22 @@
+#import <Foundation/Foundation.h>
```
Can we move this test from `objc-correctness` to `reference`?

zoecarver commented on test/Interop/Cxx/objc-correctness/const-ref-parameter.swift:
```diff
@@ -0,0 +1,77 @@
+// CHECK-IDE-TEST: mutating func doOtherThing(_ options: OptionsStruct) -> Float
+
+
+// RUN: %target-swift-frontend -c -enable-experimental-cxx-interop -enable-objc-interop -I %S/Inputs %s -emit-sil -o - | %FileCheck %s
```
Nit: let's check silgen to match all the other tests. Some early sil passes might move things around.

zoecarver replied:
For clarity, I mean `-emit-silgen`.

zoecarver commented on test/Interop/Cxx/objc-correctness/const-ref-parameter.swift:
```diff
@@ -0,0 +1,77 @@
+// CHECK-SAME: : $@convention(c) (OptionsStruct) -> @out OptionsConsumerCxx
+
+// COM: FIXME: should it be @in_guaranteed OptionStruct?
+// CHECK: [[FN6:%[0-9]+]] = function_ref @_ZN18OptionsConsumerCxx12doOtherThingERK13OptionsStruct : $@convention(cxx_method) (@in OptionsStruct, @inout OptionsConsumerCxx) -> Float
```
Is this passing on Windows?

zoecarver replied:
Oh, I see there's a requires objc at the top.

drodriguez replied:
I don't know what would change for Windows. The mangling?

As a later PR I can split the test into a pure C++ one and a Obj-C++ one, so at least part of the test runs in Linux and Windows. I don't have a Windows machine anymore, but I can check that Linux is consistent with Darwin.

zoecarver commented:
@beccadax would you mind looking over these changes (as they touch objc method importing).

drodriguez committed:
[NFC] Remove out parameters. Create struct for return type.

Instead of having out parameters for a couple of flags, create a small
struct with the type and the flags and return that struct inside the
Optional of importParameterType.

Additionally change the naming of some variables to make the function
and method code more similar.

drodriguez commented:
@swift-ci please test

drodriguez commented:
@swift-ci please test source compatibility

plotfi commented:
@beccadax @zoecarver Is there more to look at for this PR before merging?

drodriguez commented:
I am going to merge this in order to unblock other changes after this. If there's a problem, we should be able to fix forward or if there's a very big problem, reverting is also fine. Hope I do not disrupt anyone's work.

plotfi commented:
Merging @drodriguez. Will accept post commit reviews @zoecarver @beccadax. Thanks.

zoecarver commented on test/Interop/Cxx/reference/const-ref-parameter.swift:
```diff
@@ -0,0 +1,77 @@
+    _ = OptionsConsumerCxx.doThing(a)
+}
+
+// RUN: %target-swift-ide-test -print-module -module-to-print=ConstRefParameter -I %S/Inputs -source-filename=x -enable-experimental-cxx-interop -enable-objc-interop | %FileCheck -check-prefix=CHECK-IDE-TEST %s
```
This should go at the top of the file.

zoecarver commented on test/Interop/Cxx/reference/const-ref-parameter.swift:
```diff
@@ -0,0 +1,77 @@
+
+// CHECK-IDE-TEST: class OptionsConsumerObjC
+// CHECK-IDE-TEST: init(options: OptionsStruct)
+// COM: FIXME: should it be consumer(options:)?
```
I don't see why?

drodriguez replied:
According to https://github.com/apple/swift/blob/main/docs/CToSwiftNameTranslation.md#factory-initializers for a factory initializer the "with" should be dropped. I don't know why it is not happening here.

But also, there's a phrase that says "If a factory initializer turns out to have the same imported name as an available designated initializer, or a non-inherited factory initializer has the same imported name as an available convenience initializer, the factory initializer is marked unavailable to resolve ambiguities." which also should had happen here, but it doesn't (`consumer` should have turn into `init`).

Something is wonky here, but I am not sure what exactly is failing.

zoecarver commented on test/Interop/Cxx/reference/const-ref-parameter.swift:
```diff
@@ -0,0 +1,77 @@
+// CHECK-IDE-TEST: mutating func doOtherThing(_ options: OptionsStruct) -> Float
+
+
+// RUN: %target-swift-frontend -c -enable-experimental-cxx-interop -enable-objc-interop -I %S/Inputs %s -emit-silgen -o - | %FileCheck %s
```
Again this should go at the top of the file.

zoecarver commented on test/Interop/Cxx/reference/Inputs/const-ref-parameter.h:
```diff
@@ -0,0 +1,22 @@
+
+struct OptionsConsumerCxx {
+  OptionsConsumerCxx(const OptionsStruct &options);
+  static OptionsConsumerCxx build(const OptionsStruct &options);
+  static int doThing(const OptionsStruct &options);
+  float doOtherThing(const OptionsStruct &options);
```
These method names aren't particularly helpful. Maybe you can rename them "getIntOption" and "getFloatOption" or "staticMethod" and "instanceMethod" depending on why they are interesting test cases.

zoecarver commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2230,6 +2230,161 @@ ImportedType ClangImporter::Implementation::importFunctionParamsAndReturnType(
+             isa<clang::TemplateTypeParmType>(paramTy->getPointeeType())) {
+    auto templateParamType =
+        cast<clang::TemplateTypeParmType>(paramTy->getPointeeType());
+    swiftParamTy = findGenericTypeInGenericDecls(
```
It seems like this patch enables a lot of new behavior for Objective-C++ methods (for example, templated methods, if those are even allowed). Can you add some more tests? Otherwise people will use them and we'll have no idea if they work or not.

drodriguez commented on test/Interop/Cxx/reference/const-ref-parameter.swift:
```diff
@@ -0,0 +1,77 @@
+
+// RUN: %target-swift-frontend -c -enable-experimental-cxx-interop -enable-objc-interop -I %S/Inputs %s -emit-silgen -o - | %FileCheck %s
+
+// COM: FIXME: should it be @in_guaranteed OptionsStruct? https://github.com/apple/swift/issues/60601
```
Most of the comments are removed in Puyan's next PR (actually, this one is not actually removed). I will review after his PR and try to remove duplicated comments.

drodriguez reviewed (commented):
I will try to move the `RUN` lines at the top of the file after Puyan merges his PR. I did not know there's a preferred style (I prefer to have them close to the check lines, if possible).

beccadax reviewed (commented):
Retrospective review: Looks generally good. I have one concern that doesn’t seem to have been discussed yet.

drodriguez commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2904,7 +2911,7 @@ ImportedType ClangImporter::Implementation::importMethodParamsAndReturnType(
     }
 
     // Special case for NSDictionary's subscript.
-    ImportTypeKind importKind = ImportTypeKind::Parameter;
```
I think this was a difference between the `importMethodParamsAndReturnType` and `importFunctionParameterList` that were in different places, but had the same results.

(I will quote the old lines in the rest of the text)

In `importFunctionParameterList` in set in 2257-2261, almost being the first thing done in the method, but `importKind` was not used until 2326 inside the check for `!swiftParamTy`.

In `importMethodParamsAndReturnType` is set in 2907 and only overwritten in 2937-2940 inside another `!swiftParamTy` check (in this case an `else if`, not a naked `if`). The lines 2937-2940 are the same as the lines 2258-2261, and I don't think there's any change in `param` that might change the result of those lines being together with the first line or separated.

From my reading of the code it seemed that the result in `importKind` only depends on the initial (constant) value of `param`, and it was not changing over the execution of the method. That's why it was extracted into a stateless helper and move closer to its usage.

I might have not seen some detail of how variables evolve, but I think the code should still create the same results.
